Training/python/Phase2/Training.py

```diff
@@ -31,6 +31,231 @@
+        self.loss_tracker.update_state(loss_vec, sample_weight=sample_weight)
+        self.pure_loss_tracker.update_state(pure_loss_vec, sample_weight=sample_weight) 
+        self.reg_loss_tracker.update_state(reg_loss)
+        self.compiled_metrics.update_state(y, y_pred, sample_weight)
+        # Return a dict mapping metric names to current value
+        metrics_out = {m.name: m.result() for m in self.metrics}
+        self.x = x
+        self.y = y
+        self.sample_weight = sample_weight
+        self.y_pred = y_pred
+        return metrics_out
+    
+    @property
+    def metrics(self):
+        # define metrics here so that `reset_states()` can be
+        # called automatically at the start of each epoch
+        # or at the start of `evaluate()`
+        metrics = []
+        metrics.append(self.loss_tracker) 
+        metrics.append(self.reg_loss_tracker)
+        metrics.append(self.pure_loss_tracker)
+        if self._is_compiled:
+            #  Track `LossesContainer` and `MetricsContainer` objects
+            # so that attr names are not load-bearing.
+            if self.compiled_loss is not None:
+                metrics += self.compiled_loss.metrics
+            if self.compiled_metrics is not None:
+                metrics += self.compiled_metrics.metrics
+
+        for l in self._flatten_layers():
+            metrics.extend(l._metrics)  # pylint: disable=protected-access
+
+        return metrics
+
+class UpdatedLossCallback(keras.callbacks.Callback):
+    def __init__(self, data_upd=None):
+        super().__init__()
+        self.nbins = 10000
+        self.MisIDWPs = {"e": [0.065, 0.03, 0.015, 0.005, 0.002, 0.0007, 0.0003, 0.0001], "mu": [0.002, 0.0006, 0.0004, 0.0002], "jet": [0.1, 0.08, 0.04, 0.02, 0.01, 0.006, 0.004, 0.002]}
+        self.data_upd = data_upd
+
+    def on_epoch_begin(self, epoch, logs=None):
+        self.res = {}
+        self.firstbatch = True
+
+    def on_train_batch_end(self, batch, logs=None):
+        if self.firstbatch:
+            for typ,ID in zip(["e", "mu", "jet"], [0,1,3]):
+                self.res[typ] = tf.where(self.model.y[:,ID]==1.0, self.model.y_pred[:,2]/(self.model.y_pred[:,2]+self.model.y_pred[:,ID]), -1)
+                self.res["tau"+typ] = tf.where(self.model.y[:,2]==1.0, self.model.y_pred[:,2]/(self.model.y_pred[:,2]+self.model.y_pred[:,ID]), -1)
+            self.res["w"] = self.model.sample_weight
+            self.firstbatch = False
+        else:
+            for typ,ID in zip(["e", "mu", "jet"], [0,1,3]):
+                self.res[typ] = tf.concat([self.res[typ], tf.where(self.model.y[:,ID]==1.0, self.model.y_pred[:,2]/(self.model.y_pred[:,2]+self.model.y_pred[:,ID]), -1)], axis=0)
+                self.res["tau"+typ] = tf.concat([self.res["tau"+typ], tf.where(self.model.y[:,2]==1.0, self.model.y_pred[:,2]/(self.model.y_pred[:,2]+self.model.y_pred[:,ID]), -1)], axis=0)
+            self.res["w"] = tf.concat([self.res["w"], self.model.sample_weight], axis=0)
```

Comment:
Can you point which part of the code requires eager execution, it is this one?

Reply:
Yes, that's right. I save the `y` and `y_pred` from `train_step` to the class and read them here. `self.model.y`, `self.model.y_pred` and `self.model.sample_weight` are some unreadable format without eager execution.

Reply:
> Yes, that's right. I save the `y` and `y_pred` from `train_step` to the class and read them here. `self.model.y`, `self.model.y_pred` and `self.model.sample_weight` are some unreadable format without eager execution.

I see, but this format should be tensorflow tensor, and all operations here are tf.ops (since you are not using THists anymore), there should be tf-friendly way to accumulate some output values

Reply:
I would advise against using `model.sample_weight`, `model.y` etc. inside callbacks.
Another point is that you concatenate all outputs in memory, which is not scalable for big datasets. Aggregation should be used already inside `on_train_batch_end`.
I think implementing per-batch code as a metric would fit more general TF design concepts.
E.g. metric will aggregate nFakes and nTaus for each WP and then callback will process them in `on_epoch_end` (they will be available through `log` argument).

Reply:
Thanks for the idea! But I can't just save the number of Taus/Fakes passing the current WPs, I need all discriminator values to update the WPs for the next iteration.
I can do this step in a metric instead (so that doesn't require eager), but is there a way to directly access the weights inside the metric too?

Reply:
yes, `metric.update_state` takes weights as an argument. You can even call it directly if you want to have better control over how it is called. E.g. as it is done for some adversarial metrics
https://github.com/cms-tau-pog/TauMLTools/blob/ef7c8a9971a4da590907fe4af839588e32f56dee/Training/python/2018v1/Training_v0p1.py#L133

Reply:
Actually it seems that moving this implementation to a metric doesn't work either. The code for the metric is only called once before the epoch, not after every batch (At least, that's what it seems like based on some printouts). So the tensors I try to fill end up being `Tensor("SelectV2:0", shape=(None,), dtype=float32)` after the epoch.

Reply:
it is correct - there should be only one (ok, for some cases two) printout using standard python `print` when the graph is built, and then the graph is executed for each batch (no more python code calls). To confirm it you can put printout in the `train_step`. To have a printout at every real call, [tf.print](https://www.tensorflow.org/api_docs/python/tf/print) should be used.

You'll need to define your custom metric (see example [here](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/Metric)) and declare all variables to keep n_fake, n_true counts etc with `self.add_weight`. Then in `update_state` update these variables with `assign_add` or similar methods.

Reply:
I found some (old?) code on such an implementation:
https://github.com/cms-tau-pog/TauMLTools/blob/master/Training/python/commonReco.py#L58-L81
And I tried setting up the metric like this. But I seem unable to define 1-dim tensor variables to store the discriminators. There is always some error about shape mismatch, either like `ValueError: Shape must be rank 1 but is rank 0 for '{{node Reshape}} = Reshape[T=DT_FLOAT, Tshape=DT_INT32](SelectV2, Reshape/shape)' with input shapes: [?], []` or like `ValueError: Cannot assign value to variable ' TauDiscr0:0': Shape mismatch.The variable shape (1,), and the assigned value shape () are incompatible.`
I'm also not sure if this implementation would work either, because it looks like `assign_add` only adds new values on top of exists one, instead of appending them?